node/src/Pos/Explorer/DB.hs

```diff
@@ -104,10 +133,116 @@ prepareExplorerDB = do
+        epochsSource = dbIterSource GStateDB (Proxy @EpochsIter)
+
+        -- 'Conduit' to turn the old @Epoch@ format that contained all
+        -- 21600 @[HeaderHash]@ in a single @Epoch@ to a new format that contains
+        -- @Epoch@ *and* @Page@ as a key since there are too many @[HeaderHash]@ to
+        -- load - performance reasons.
+        epochPagesConduit
+            :: (MonadDBRead m)
+            => Conduit (Epoch, [HeaderHash]) m (Map EpochPagedBlocksKey [HeaderHash])
+        epochPagesConduit = CL.map convertToPagedMap
+          where
+            convertToPagedMap
+                :: (Epoch, [HeaderHash])
+                -> Map EpochPagedBlocksKey [HeaderHash]
+            convertToPagedMap ehh = fromListWith (++) (convertToPaged ehh)
+
+            convertToPaged :: (Epoch, [HeaderHash]) -> [((Epoch, Page), [HeaderHash])]
+            convertToPaged (epoch, headerHashes) =
+                convertHHsToEpochPages <$> convertHHsToPages
+              where
+                convertHHsToEpochPages
+                    :: (Page, [HeaderHash])
+                    -> ((Epoch, Page), [HeaderHash])
+                convertHHsToEpochPages (page, headerHashes') =
+                    ((epoch, page), headerHashes')
+
+                -- | Take that huge chunk of @HeaderHash@es and page it.
+                convertHHsToPages :: [(Page, [HeaderHash])]
+                convertHHsToPages = zip [1..] $ splitEvery pageSize headerHashes
+                  where
+                    -- | TODO (ks): It's getting repeated all over the place.
+                    pageSize :: Int
+                    pageSize = 10
+
+                    splitEvery :: Int -> [a] -> [[a]]
+                    splitEvery _ [] = []
+                    splitEvery n xs = as : splitEvery n bs
+                      where
+                        (as,bs) = splitAt n xs
+
+        -- | Finally, we persist the map with the new format.
+        epochPagesExplorerOpConduit
+            :: (Monad m)
+            => Conduit (Map EpochPagedBlocksKey [HeaderHash]) m [ExplorerOp]
+        epochPagesExplorerOpConduit = CL.map persistEpochBlocks
+          where
+            -- | Persist atomically, all the operations together.
+            persistEpochBlocks :: Map EpochPagedBlocksKey [HeaderHash] -> [ExplorerOp]
+            persistEpochBlocks mapEpochPagedHHs =
+                persistEpochPageBlocks ++ persistMaxPageNumbers
+              where
+                -- | Persist @Epoch@ @Page@ blocks.
+                persistEpochPageBlocks :: [ExplorerOp]
+                persistEpochPageBlocks = putKeyBlocks <$> M.toList mapEpochPagedHHs
+                  where
+                    putKeyBlocks
+                        :: (EpochPagedBlocksKey, [HeaderHash])
+                        -> ExplorerOp
+                    putKeyBlocks keyBlocks = PutEpochBlocks epoch page blocks
+                      where
```

Comment:
I believe having 4-5 nested `where` clauses makes the code much less readable. Please avoid deep nesting.

Reply:
Deep nesting is indicating local calls. I prefer this than to pollute the namespace - this is where polluting the namespace is actually a consequence of removing/flattening the clauses.

Reply:
For now, keep it as-is, but we should decide on whether this is acceptable style or not (a quick survey showed varying results) and stick to the style guide.